### PR TITLE
feat: unify weaverbird pipeline failure exceptions (TCTC-9178)

### PR DESCRIPTION
**Pull request by julien-pinchelimouroux**

### CODE REVIEW

File: server/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+* Error management improvement: 
+  - Add a global `PipelineFailure` exception to make easier to catch translation and execution errors from supported backends.
+  - Now, on error, Pypika backends raise a `PipelineTranslationFailure` which specify the error, and also the step name and the step index if known.
```
Comment (lukapeschke) [inline server/CHANGELOG.md:7]:
```suggestion
  - A `PipelineFailure` exception was added to make it easier to catch translation and execution errors from supported backends. It is the parent excetpion of both `PipelineTranslationFailure` and `PipelineExecutionFailure`
  - On error, Pypika backends now raise a `PipelineTranslationFailure` which specifies the error, the step name and the step index if known. 
```

File: server/src/weaverbird/backends/pypika_translator/translators/base.py [hunk 2]
```diff
@@ -337,9 +340,11 @@ def get_query_builder(
...
         unwrap_last_step: bool = False,
     ) -> QueryBuilderContext:
-        if len(steps) < 0:
-            ValueError("No steps provided")
-        assert steps[0].name == "domain" or steps[0].name == "customsql"
+        if len(steps) <= 0:
+            raise ValueError("No steps provided")
+        assert (
+            steps[0].name == "domain" or steps[0].name == "customsql"
+        ), "First step must be one of domain or customsql step"
```
Comment (lukapeschke) [inline server/src/weaverbird/backends/pypika_translator/translators/base.py:347]: Nice, could you please also add the step name in the assertion message ?

**Review by lukapeschke — APPROVED** [review]
Thanks!

File: server/src/weaverbird/backends/pypika_translator/translators/base.py (hunk 2 above)
Comment (julien-pinchelimouroux) [reply server/src/weaverbird/backends/pypika_translator/translators/base.py]: [75b5cc8](https://github.com/ToucanToco/weaverbird/pull/2238/commits/75b5cc85f88f163b6c8e32cd51fdaf1f56ff4c56)

File: server/CHANGELOG.md (hunk 1 above)
Comment (julien-pinchelimouroux) [reply server/CHANGELOG.md]: [75b5cc8](https://github.com/ToucanToco/weaverbird/pull/2238/commits/75b5cc85f88f163b6c8e32cd51fdaf1f56ff4c56)

**Pull request merged**